prov/util/src/util_ns.c

```diff
@@ -312,23 +313,21 @@ static void *util_ns_name_server_func(void *args)
 	if (ret)
 		goto done;
 
-	cleanup_args[0] = (void *)(uintptr_t)listenfd;
+	cleanup_args[0] = &listenfd;
 	cleanup_args[1] = (void *)ns;
 	pthread_cleanup_push(util_ns_name_server_cleanup,
 			     (void *)cleanup_args);
-
 	while (1) {
-		connfd = accept(listenfd, NULL, 0);
-		if (connfd != INVALID_SOCKET) {
-			/* Read service data */
-			ret = util_ns_read_socket_op(connfd, &cmd, cmd_len);
-			if (!ret) {
-				(void) util_ns_op_dispatcher(ns, &cmd, connfd);
-			}
-			ofi_close_socket(connfd);
-		}
+		if (listenfd != INVALID_SOCKET)
```

Comment:
This isn't the right place for this check.  If listenfd is invalid, we shouldn't be in this loop.  I'm guessing there are already checks in place for this.

The original code was checking connfd.  If that returns an invalid socket, then we likely have some sort of error with the listen and should break the loop.

Reply:
I think the logic here is that listenfd can be set to invalid asynchronously by the cleanup routine when the thread is cancelled.

Reply:
Hmm... so another thread is closing the listening socket then?  I don't see that in the code.  But this sounds racy between closing the listening socket and setting the variable to invalid.  So, we can't rely on listen being invalid to kick us out of the loop (at least immediately).  It seems that we should always exit the loop if accept() returns anything other than a usable socket.

The structure of the code looks more complex than is needed.  To stop the server, we should just need to close the listening socket and wait for the thread to exit.  I'm particularly concerned that we have a call -- ofi_osd_fini() -- from one thread while another thread is actively using the networking API.

Reply:
Here is the related code (also in this patch):
```
static void util_ns_name_server_cleanup(void *args)  
180 180    {  
181 181    	void **cleanup_args = (void **)args;  
182   -	 	ofi_close_socket((uintptr_t)cleanup_args[0]);  
182   +	 	ofi_close_socket(*((SOCKET *)cleanup_args[0]));  
183   +	 	*((SOCKET *)cleanup_args[0]) = INVALID_SOCKET;  
183 184    	util_ns_map_fini((struct util_ns *)cleanup_args[1]);  
184 185    }  
```
It used to be closing the fd directly. The patch changes it to also modify the fd variable to INVALID_SOCKET.

Reply:
I'm suggesting that we look at removing the pthread_cleanup_push/pop calls, and rework the code.  I would move most of the code from util_ns_name_server_func() into ofi_ns_start_server().  That is, start_server() should start listening for connections.  Then we spawn a thread to do this:
```
util_ns_conn_handler(...)
{
    while (listenfd != INVALID_SOCKET) {
        connfd = accept(...);
        if (connfd == INVALID_SOCKET)
            break;
        ...
    };
}
```

The stop_server routine then does:
```
    ofi_close_socket(listenfd);
    listenfd = INVALID_SOCKET; /* not actually needed */
    pthread_join(...);
    ofi_osd_fini();
```